### ISSUE TITLE
Less type inference with fold_many0 on nom5 

### ISSUE THREAD

**Issue by alreece45** [Geal/nom#962]
A named parser like this compiles on nom4 and nom3, but fails to compile on nom5-- without a clear path to resolve the error:

```rust
named!(multi<&[u8], Vec<&[u8]>>,
    fold_many0!(
        alt!(tag!("aaaa") | tag!("bbbb")),
        Vec::new(), |mut acc: Vec<_>, item| {
            acc.push(item);
            acc
        }
    )
);
```

```
error[E0282]: type annotations needed
  --> nom5-sandbox/src/main.rs:13:1
   |
13 | / named!(multi<&[u8], Vec<&[u8]>>,
14 | |     fold_many0!(
15 | |         alt!(tag!("abcd") | tag!("defg")),
16 | |         Vec::new(), |mut acc: Vec<_>, item| {
...  |
20 | |     )
21 | | );
   | |__^ consider giving this closure parameter a type
   |
   = note: type must be known at this point
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Even adding type parameters and using static lifetimes seem to be unable to resolve the error:

```rust
named!(multi<&[u8], Vec<&[u8]>>,
    fold_many0!(
        alt!(tag!("aaaa") | tag!("bbbb")),
        Vec::new(), |mut acc: Vec<&'static [u8]>, item: &[u8]| -> Vec<&'static [u8]> {
            acc.push(&b"123"[..]);
            acc
        }
    )
);
```

```error[E0282]: type annotations needed
  --> nom5-sandbox/src/main.rs:13:1
   |
13 | / named!(multi<&[u8], Vec<&[u8]>>,
14 | |     fold_many0!(
15 | |         alt!(tag!("aaaa") | tag!("bbbb")),
16 | |         Vec::new(), |mut acc: Vec<&'static [u8]>, item: &[u8]| -> Vec<&'static [u8]> {
...  |
20 | |     )
21 | | );
   | |__^ consider giving this closure parameter a type
   |
   = note: type must be known at this point
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)```

**Comment by alreece45** [comment]
Pulling the `alt!` into its own named function allows it to compile:

```named!(multi<&[u8], Vec<&[u8]>>,
    fold_many0!(
        call!(multi_part),
        Vec::new(), |mut acc: Vec<_>, item: &[u8]| -> Vec<&[u8]> {
            acc.push(&b"123"[..]);
            acc
        }
    )
);

named!(multi_part<&[u8]>,
    alt!(tag!("aaaa") | tag!("bbbb"))
);

```

**Comment by Geal** [comment]
maybe this one can be solved with a `fold_many0c` function like other combinators

**Comment by eggyal** [comment]
The `fold_many0c` function was added by fa63ba4, which it seems has been in v5.0.0 since the start—but the `fold_many0!` macro doesn't call to it.  Simply making this change fixed it for me:

```
diff --git a/src/multi/macros.rs b/src/multi/macros.rs
index 10b66b8..8565a2a 100644
--- a/src/multi/macros.rs
+++ b/src/multi/macros.rs
@@ -450,7 +450,7 @@ macro_rules! fold_many0(
     fold_many0!($i, |i| $submac!(i, $($args)*), $init, $fold_f)
   );
   ($i:expr, $f:expr, $init:expr, $fold_f:expr) => (
-    $crate::multi::fold_many0($f, $init, $fold_f)($i)
+    $crate::multi::fold_many0c($i, $f, $init, $fold_f)
   );
 );
```

**Comment by Geal** [comment]
fixed in 6168f7b
